odoo/fields.py

```diff
@@ -3315,6 +3316,77 @@ def convert_to_onchange(self, value, record, names):
         self._add_display_name(value, record.env)
         return value
 
+    def read(self, records):
+        """Read everything needed in batch for the given records.
+
+        To retrieve relational properties names, or to check their existence,
+        we need to do some SQL queries. To reduce the number of queries when we read
+        in batch, we put in cache everything needed before calling
+        convert_to_record / convert_to_read.
+        """
+        definition_records_map = {
+            record: record[self.definition_record][self.definition_record_field]
+            for record in records
+        }
+
+        # ids per model we need to fetch in batch to put in cache
+        ids_per_model = defaultdict(list)
+
+        cached_values = list(records.env.cache.get_values(records, self))
+
+        for record, values in zip(records, cached_values):
+            definition = definition_records_map.get(record)
+            if not values or not definition:
+                continue
+            for property_definition in definition:
+                comodel = property_definition.get('comodel')
+                type_ = property_definition.get('type')
+                name = property_definition.get('name')
+                if not comodel or type_ not in ('many2one', 'many2many') or name not in values:
+                    continue
+
+                default = property_definition.get('default') or []
+                value = values[name] or []
+
+                if type_ == 'many2one':
+                    default = [default] if default else []
+                    value = [value] if value else []
+
+                ids_per_model[comodel].extend(default + value)
```

Comment:
You add the default also when the value is set, why ?

Also a record id can be added several times, you should use a OrderedSet (instead of list) to avoid duplication.

Reply:
> You add the default also when the value is set, why ?

Because we might also need the display name of the default value (and check their existence as well

> Also a record id can be added several times, you should use a OrderedSet (instead of list) to avoid duplication.

Good catch :) I can even use a set because order does not matter here

Reply:
> Good catch :) I can even use a set because order does not matter here

Unfortunately, yes it does matter. Because you make a recordset and call a compute (sometime) field (display_name) and the order can change the number of query generated (we don't want that). Then you should use a OrderedSet :/